### PR TITLE
[FLINK-19650][connectors jdbc]Support the limit push down for the Jdb…

### PR DESCRIPTION
**Pull request by fsk119**
…c connector

## What is the purpose of the change

*Support the limit push down for the jdbc connectors*

## Brief change log

  - *The jdbc connector extends the `SupportsLimitPushDown` interface*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added IT case to verify the results.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java [hunk 1]
```diff
@@ -92,8 +99,14 @@ public ScanRuntimeProvider getScanRuntimeProvider(ScanContext runtimeProviderCon
...
-		if (readOptions.getFetchSize() != 0) {
-			builder.setFetchSize(readOptions.getFetchSize());
+		if (readOptions.getFetchSize() != 0 || this.limit >= 0) {
+			int fetchsize = readOptions.getFetchSize();
+			if (fetchsize == 0) {
+				fetchsize = limit;
+			} else if (limit != -1) {
+				fetchsize = Integer.min(fetchsize, limit);
+			}
+			builder.setFetchSize(fetchsize);
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:109]: We don't need to adapt fetch size, it is a best effort read size.

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java [hunk 2]
```diff
@@ -154,7 +155,7 @@ public void testProject() throws Exception {
 				")"
 		);
 
-		Iterator<Row> collected = tEnv.executeSql("SELECT id,timestamp6_col,decimal_col FROM " + INPUT_TABLE).collect();
+		Iterator<Row> collected = tEnv.executeSql("SELECT id,timestamp6_col,decimal_col FROM " + INPUT_TABLE + " LIMIT 1").collect();
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java:158]: This doesn't break the test?

File: flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java [hunk 3]
```diff
@@ -166,4 +167,38 @@ public void testProject() throws Exception {
...
+						"double_col DOUBLE," +
+						"decimal_col DECIMAL(10, 4)" +
+						") WITH (" +
+						"  'connector'='jdbc'," +
+						"  'url'='" + DB_URL + "'," +
+						"  'table-name'='" + INPUT_TABLE + "'" +
+						")"
+		);
+
+		Iterator<Row> collected = tEnv.executeSql("SELECT id FROM " + INPUT_TABLE + " LIMIT 1").collect();
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/test/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSourceITCase.java:196]: Please select all the fields.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java [hunk 4]
```diff
@@ -156,11 +169,25 @@ public boolean equals(Object o) {
...
+	 * {@link java.sql.Statement#setFetchSize(int)} only accepts int value.
+	 */
+	@Override
+	public void applyLimit(long limit) {
+		if (limit > Integer.MAX_VALUE) {
+			throw new TableException(
+					String.format("The maximum limit value is %d for jdbc connector. Get %d.",
+							Integer.MAX_VALUE, limit));
+		}
+		this.limit = Math.toIntExact(limit);
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:191]: We don't need to cast to int.

**Review by wuchong — CHANGES_REQUESTED** [review]
Thanks for the contribution, I left some comments.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java [hunk 5]
```diff
@@ -148,14 +150,15 @@ default String getDeleteStatement(String tableName, String[] conditionFields) {
...
 		String selectExpressions = Arrays.stream(selectFields)
 				.map(this::quoteIdentifier)
 				.collect(Collectors.joining(", "));
 		String fieldExpressions = Arrays.stream(conditionFields)
 				.map(f -> format("%s = :%s", quoteIdentifier(f), f))
 				.collect(Collectors.joining(" AND "));
 		return "SELECT " + selectExpressions + " FROM " +
-				quoteIdentifier(tableName) + (conditionFields.length > 0 ? " WHERE " + fieldExpressions : "");
+				quoteIdentifier(tableName) + (conditionFields.length > 0 ? " WHERE " + fieldExpressions : "") +
+				(limit >= 0 ? " " + getLimit(limit) : "");
```
Comment (leonardBang) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:162]: move `limit >= 0 ? " " + getLimit(limit) : ""` to  `getLimit(long limit)` internal?

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java [hunk 6]
```diff
@@ -156,11 +163,17 @@ public boolean equals(Object o) {
...
 			Objects.equals(lookupOptions, that.lookupOptions) &&
 			Objects.equals(physicalSchema, that.physicalSchema) &&
-			Objects.equals(dialectName, that.dialectName);
+			Objects.equals(dialectName, that.dialectName) &&
+			Objects.equals(limit, that.limit);
 	}
 
 	@Override
 	public int hashCode() {
 		return Objects.hash(options, readOptions, lookupOptions, physicalSchema, dialectName);
```
Comment (leonardBang) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/table/JdbcDynamicTableSource.java:172]: please also add missed `limit `

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala [hunk 7]
```diff
@@ -244,6 +244,7 @@ object FlinkStreamRuleSets {
     PushProjectIntoLegacyTableSourceScanRule.INSTANCE,
     PushFilterIntoTableSourceScanRule.INSTANCE,
     PushFilterIntoLegacyTableSourceScanRule.INSTANCE,
+    PushLimitIntoTableSourceScanRule.INSTANCE,
```
Comment (leonardBang) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala:247]: I think this rule should have been existed after support limit pushdown in planner, why we add it util now?

**Review by leonardBang — CHANGES_REQUESTED** [review]
Thanks @fsk119 for the contribution, I left several comments

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala (hunk 7 above)
Comment (fsk119) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/FlinkStreamRuleSets.scala:247]: They only add the rule into `BatchRuleSet`. I add a new commit to test the rule under the StreamEnvironment.

File: flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/LimitableSourceTest.scala [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.stream.sql
+
+import org.junit.Before
+
+/**
+ * Test for.
```
Comment (leonardBang) [inline flink-table/flink-table-planner-blink/src/test/scala/org/apache/flink/table/planner/plan/stream/sql/LimitableSourceTest.scala:24]: for what ? ^_^

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/MySQLDialect.java [hunk 9]
```diff
@@ -55,6 +55,15 @@ public JdbcRowConverter getRowConverter(RowType rowType) {
 		return new MySQLRowConverter(rowType);
 	}
 
+	@Override
+	public String getLimit(long limit) {
```
Comment (leonardBang) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/MySQLDialect.java:59]: how about rename to `getLimitStatement` to make the function definition more clear?

**Review by leonardBang — APPROVED** [review]
Thanks @fsk119 for the update, +1 from my side

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java [hunk 10]
```diff
@@ -148,14 +150,15 @@ default String getDeleteStatement(String tableName, String[] conditionFields) {
...
 		String selectExpressions = Arrays.stream(selectFields)
 				.map(this::quoteIdentifier)
 				.collect(Collectors.joining(", "));
 		String fieldExpressions = Arrays.stream(conditionFields)
 				.map(f -> format("%s = :%s", quoteIdentifier(f), f))
 				.collect(Collectors.joining(" AND "));
 		return "SELECT " + selectExpressions + " FROM " +
-				quoteIdentifier(tableName) + (conditionFields.length > 0 ? " WHERE " + fieldExpressions : "");
+				quoteIdentifier(tableName) + (conditionFields.length > 0 ? " WHERE " + fieldExpressions : "") +
+				getLimitStatement(limit);
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:162]: Personally, I don't like the current design of the new interface. 
 - it couples the `limit` to the select statement which makes it impossible to inject additional where clause. 
 - dialects need to implement/take care 2 interfaces about limit. 
 - many invokers have to pass an invalid limit. 

In my opinion, a better solution would be separate them, by having a new interface for limit:

```java
String getLimitClause(long limit);
```

And concat the limit caluse with the select statement where needs the limit clause.

File: flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java [hunk 11]
```diff
@@ -57,6 +57,8 @@
 	 */
 	JdbcRowConverter getRowConverter(RowType rowType);
 
+	String getLimitStatement(long limit);
```
Comment (wuchong) [inline flink-connectors/flink-connector-jdbc/src/main/java/org/apache/flink/connector/jdbc/dialect/JdbcDialect.java:60]: Please add javadoc on the method, including the description for limit parameter, e.g. the value range.

**Review by wuchong — CHANGES_REQUESTED** [review]
Thanks @fsk119 for the great work!  I left some suggestions.

**Review by wuchong — APPROVED** [review]
LGTM.

**Pull request merged**